### ISSUE TITLE
2.1.0 doubles instances when using .restart

### ISSUE THREAD

**Issue by Gremagol** [Kwoth/NadekoBot#1743]
As the title says, everytime you restart the bot, it doubles the amount of instances. To understand this look at the images.
How to reproduce: update to 2.1.0, type .restart, enjoy
![bug8](https://user-images.githubusercontent.com/24619854/31907235-880503d6-b833-11e7-87fc-36009b156522.png)

![bug10](https://user-images.githubusercontent.com/24619854/31907245-8fef5038-b833-11e7-8611-225d6d1e80be.PNG)

![bug9](https://user-images.githubusercontent.com/24619854/31907242-8c9dd15c-b833-11e7-8d02-d7ee018b4814.png)

Also, check the console while restarting. The logs doubled too:
![bug11](https://user-images.githubusercontent.com/24619854/31907403-00e240f2-b834-11e7-83a8-fd2e1d411a12.PNG)

**Comment by ManuelFte** [comment]
I did a clean install in a new folder and saving only credentials and database, the error happened without even using `.restart`. Also I noticed that when the bot was launching the first time, it printed this:

```
13:49:01 ShardsCoordinator | Shard 0 is scheduled for a restart because it's unresponsive.
13:49:01 ShardsCoordinator | Auto-restarting shard 0
```

It looks like it's detecting the shards as unresponsive and tries to autorestart them when they haven't been started even once yet.

**Comment by ManuelFte** [comment]
Another error I noticed when I use `.restart`: the process gets stuck and the RAM usage increases non-stop until exhausting the resources of the machine and freezing it:

![screenshot from 2017-10-23 14-14-16](https://user-images.githubusercontent.com/896973/31908662-3e339a96-b7fd-11e7-84fe-a76a5f60bea8.png)

**Comment by ManuelFte** [comment]
Probably related issue: 2.0.4 and 2.1.0 have a very high CPU usage. At some points one of the bots started using 96% (a single core at 2.4 GHz while being idle. Restarting the bot didn't solve anything. Only when downgrading to 2.0.3 the CPU usage decreased to normal levels (around 1% and 2%).

**Comment by shikhir-arora** [comment]
It is related. It's sort of similar to https://github.com/Kwoth/NadekoBot/issues/1736 
As in that bug, one of the issue was this: https://user-images.githubusercontent.com/8797918/31860026-51f725a0-b6e2-11e7-8d3d-c5315e72393d.png

The same thing happens here because the ShardCoordinator restarts the shard 0 right on startup and spawns an extra thread. That's also why you're seeing high CPU which I can also confirm.

The second thread isn't actually 'active' but eats up resources and is locked. Only difference is with 2.1.0 since there's a auto-restart, killing the process manually doesn't fix it, it'll instantly trigger the bot to restart (since it's shard 0) and spawn it. You can confirm by seeing the secondary process will have the parent process ID and shard number as part of the name, as shown in that pic. 

Here's another one from 2.1.0:  
<img width="1078" alt="screen shot 2017-10-23 at 4 33 19 pm" src="https://user-images.githubusercontent.com/8797918/31912004-848c17c6-b810-11e7-8d54-36d77d2e70d9.png">

**Comment by shikhir-arora** [comment]
https://github.com/Kwoth/NadekoBot/blob/fb6d89368e52f6b49a5dadd34c496814cee10b0d/NadekoBot.Core/Services/ShardsCoordinator.cs#L232 is certainly the reason the CPU usage goes to 100%, and there's a ton of threads spawned for the processes - really all of them in the child process that are asleep but spawned due to one of the loops I'd assume, see pic). Putting a delay interval instead of just`while (true);`  should be an easy fix, but likely isn't the most efficient fix either.

<img width="1096" alt="screen shot 2017-10-23 at 9 56 25 pm" src="https://user-images.githubusercontent.com/8797918/31921641-73090468-b83e-11e7-9bef-cf0f711b776a.png">

**Comment by Kwoth** [comment]
Do not use .restart if your bot is on auto-restart option. Use `.die`

**Comment by Kwoth** [comment]
Oh, I see, actually this is because of the new shard coordinator.

**Comment by Kwoth** [comment]
Kill the bot completely, check in the running tasks, and then update and start the bot.

**Comment by shikhir-arora** [comment]
The CPU usage seems to be the main thing though. The restart wasn’t an issue for me, it made sense why that doubled the instances because the json specifies dotnet exec again. But the new ShardsCoordinator infinite loop is blocking the whole CPU. I have a 8 core system so it doesn’t crash as it’s just one core that’s occupied, but a single-core system would certainly get there. And the actual coordinator does work; killing any thread or lightweight thread associated with it will cause it to restart. But it’s likely that while (true); that’s causing it to block to 100%.

**Comment by shikhir-arora** [comment]
And yeah there are no threads when I stop the bot, or any running dotnet processes.

**Comment by Kwoth** [comment]
Fixed cpu usage too in 2.1.2

**Comment by shikhir-arora** [comment]
Yeah it's all good, thanks 👍
